File: framework/src/play/src/main/scala/play/api/libs/typedmap/TypedKey.scala
/*
 * Copyright (C) 2009-2018 Lightbend Inc. <https://www.lightbend.com>
 */

package play.api.libs.typedmap

/**
 * A TypedKey is a key that can be used to get and set values in a
 * [[TypedMap]] or any object with typed keys. This class uses reference
 * equality for comparisons, so each new instance is different key.
 *
 * @param displayName The name to display for this key or `null` if
 * no display name has been provided. This name is only used for debugging.
 * Keys with the same name are not considered to be equal.
 * @tparam A The type of values associated with this key.
 */
final class TypedKey[A] private (val displayName: Option[String]) {

  /**
   * Bind this key to a value. This is equivalent to the `->` operator.
   *
   * @param value The value to bind this key to.
   * @return A bound value.
   */
  def bindValue(value: A): TypedEntry[A] = TypedEntry(this, value)

  /**
   * Bind this key to a value. Equivalent to [[bindValue]].
   *
   * @param value The value to bind.
   * @return An entry binding this key to a value of the right type.
   */
  def ->(value: A): TypedEntry[A] = bindValue(value)

  override def toString: String = displayName.getOrElse(super.toString)

  /**
   * @return The Java version for this key.
   */
  def asJava: play.libs.typedmap.TypedKey[A] = new play.libs.typedmap.TypedKey[A](this)
}

/**
 * Helper for working with `TypedKey`s.
 */
object TypedKey {

  /**
   * Creates a [[TypedKey]] without a name.
   *
   * @tparam A The type of value this key is associated with.
   * @return A fresh key.
   */
  def apply[A]: TypedKey[A] = new TypedKey[A](None)

  /**
   * Creates a [[TypedKey]] with the given name.
   *
   * @param displayName The name to display when printing this key.
   * @tparam A The type of value this key is associated with.
   * @return A fresh key.
   */
  def apply[A](displayName: String): TypedKey[A] = new TypedKey[A](Some(displayName))
}

File: framework/src/play/src/main/java/play/libs/typedmap/TypedKey.java
/*
 * Copyright (C) 2009-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package play.libs.typedmap;

import play.api.libs.typedmap.TypedKey$;

/**
 * A TypedKey is a key that can be used to get and set values in a
 * {@link TypedMap} or any object with typed keys. This class uses reference
 * equality for comparisons, so each new instance is different key.
 */
public final class TypedKey<A> {

    private final play.api.libs.typedmap.TypedKey<A> underlying;

    public TypedKey(play.api.libs.typedmap.TypedKey<A> underlying) {
        this.underlying = underlying;
    }

    /**
     * @return the underlying Scala TypedKey which this instance wraps.
     *
     * @deprecated As of release 2.6.8. Use {@link #asScala()}
     */
    @Deprecated
    public play.api.libs.typedmap.TypedKey<A> underlying() {
        return underlying;
    }

    /**
     * @return the underlying Scala TypedKey which this instance wraps.
     */
    public play.api.libs.typedmap.TypedKey<A> asScala() {
        return underlying;
    }

    /**
     * Bind this key to a value.
     *
     * @param value The value to bind this key to.
     * @return A bound value.
     */
    public TypedEntry<A> bindValue(A value) {
        return new TypedEntry<A>(this, value);
    }

    /**
     * Creates a TypedKey without a name.
     *
     * @param <A> The type of value this key is associated with.
     * @return A fresh key.
     */
    public static <A> TypedKey<A> create() {
        return new TypedKey<>(TypedKey$.MODULE$.apply());
    }

    /**
     * Creates a TypedKey with the given name.
     *
     * @param displayName The name to display when printing this key.
     * @param <A> The type of value this key is associated with.
     * @return A fresh key.
     */
    public static <A> TypedKey<A> create(String displayName) {
        return new TypedKey<>(TypedKey$.MODULE$.apply(displayName));
    }

    @Override
    public String toString() {
        return underlying.toString();
    }

    @Override
    public int hashCode() {
        return underlying.hashCode();
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof TypedKey) {
            return this.underlying.equals(((TypedKey) obj).underlying);
        } else {
            return false;
        }
    }
}
